docs/RFCS/20220818_tenant_version_upgrades.md

```diff
@@ -0,0 +1,475 @@
+**Table of Contents**
+
+- [Summary](#summary)
+- [Background](#background)
+    - [Introduction](#introduction)
+    - [Invariant between logical and executable versions](#invariant-between-logical-and-executable-versions)
+    - [What gets controlled by logical versions](#what-gets-controlled-by-logical-versions)
+    - [Additional invariants](#additional-invariants)
+    - [Where the logical version is persisted](#where-the-logical-version-is-persisted)
+    - [Enforcing the invariants during upgrades and migrations in single-tenancy](#enforcing-the-invariants-during-upgrades-and-migrations-in-single-tenancy)
+    - [Complementary+legacy mechanism: startup migrations](#complementarylegacy-mechanism-startup-migrations)
+    - [How these mechanisms have been extended with multi-tenancy](#how-these-mechanisms-have-been-extended-with-multi-tenancy)
+        - [Startup migrations](#startup-migrations)
+        - [New concept: separate logical versions for tenants and the KV layer](#new-concept-separate-logical-versions-for-tenants-and-the-kv-layer)
+        - [Invariants in a multi-tenancy world](#invariants-in-a-multi-tenancy-world)
+    - [Further reading: previous RFC work](#further-reading-previous-rfc-work)
+- [Motivation](#motivation)
+    - [Why we need to do this even in a shared-process world](#why-we-need-to-do-this-even-in-a-shared-process-world)
+- [Technical design](#technical-design)
+    - [SQL server startup guardrails](#sql-server-startup-guardrails)
+    - [Version ratcheting guardrails](#version-ratcheting-guardrails)
+    - [Cleanup: remove startup migrations](#cleanup-remove-startup-migrations)
+    - [Drawbacks](#drawbacks)
+    - [Rationale and Alternatives](#rationale-and-alternatives)
+- [Explain it to folk outside of your team](#explain-it-to-folk-outside-of-your-team)
+- [Unresolved questions](#unresolved-questions)
+
+<!-- markdown-toc end -->
+
+
+# Background
+
+The following section explains how versions and upgrades work, for
+readers not yet familiar with the mechanisms. It only covers just
+enough of the current infrastructure to support the motivation section
+that follows and the technical proposal.
+
+We will move this section to a technical note (in the docs/tech-notes
+directory) once we're happy with the RFC draft.
+
+**Note: the background section is large, but the proposal/solution is very modest.**
+
+## Introduction
+
+CockroachDB uses a *logical cluster version* to organize the reveal of
+new features to users.
+
+The logical version is different from the *executable version* (the
+version of the `cockroach` program executable) as a courtesy to our
+users: it makes it possible for them to upgrade their executable
+version without stopping their cluster all at once, and organize
+changes to their SQL client apps separately.
+
+## Invariant between logical and executable versions
+
+The logical version is a number of the form 'X.Y-VVV', where X.Y is
```

Comment:
these are not quite as definitionally linked as they're presented here; you can have logical versions v21.2-z, v22.1-z, and v22.2-z etc all in the same executable version if you want, and we choose what the min version is; it currently just so happens that we pick "the one before" (some of the time), but I would not cement that fact here (we are actively working on changing this as well).

I would make a weaker but more accurate statement here, that reflects this: 

Cluster versions are just labeled points on a line. The fact the labels look like "v22.1-54" is immaterial.  There is a relationship to executables but we'll get to that in a moment. But to finish describing the line itself: the important invariants are a) we only ever move in the increasing direction on the line, b) we never move to point n+1 on the line until everyone agrees we are at n, and c) we can run code to move from n to n+1.

On the relationship to the executable version: any given executable only knows about some span of the larger line: it has an earliest point it knows about on the line, the easiest cluster version it can operate at aka its "minBinaryVersion", and a maximum point at which it can operate at called its "binaryVersion".